### PR TITLE
PXD-1164 Feat/s3

### PR DESCRIPTION
**Pull request by giangbui**

### CODE REVIEW

**Comment by rudyardrichter** [comment]
Since the idea here is to implement some basic solution for persistence: at this stage I think it would make sense to add an endpoint which forces arborist to "commit" its state to S3, so for instance fence after running a usersync can confirm that arborist will now be using that information from now on.

File: arborist/server/router_model.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+	})
+}
+
+// syncModelRouter attaches the handlers defined in this file to a main router,
+// using the prefix `/model`.
+func (server *Server) syncModelRouter(mainRouter *mux.Router) {
```
Comment (rudyardrichter) [inline arborist/server/router_model.go]: can we keep the naming consistent with the other ones and name this `addModelRouter`?

File: arborist/resource_json.go [hunk 2]
```diff
@@ -58,7 +58,8 @@ func (resourceJSON *ResourceJSON) UnmarshalJSON(data []byte) error {
 	if err != nil {
 		return err
 	}
-	resourceJSON.Path = ""
+
+	//resourceJSON.Path = ""
```
Comment (rudyardrichter) [inline arborist/resource_json.go]: just delete this line

File: arborist/resource.go [hunk 3]
```diff
@@ -103,7 +103,8 @@ func NewResource(
 		}
 		pathSegments[len(parent.pathSegments)] = name
 	} else {
-		pathSegments = []string{name}
+		//pathSegments = []string{name}
```
Comment (rudyardrichter) [inline arborist/resource.go]: delete the comment line

File: arborist/aws_client.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	awsSecretAccessKey string
+}
+
+type AwsClient struct {
+	credentials S3Credentials
+	session     *session.Session
+	bucket      string
+}
+
+func (client *AwsClient) SetS3Credentials(cred S3Credentials) {
```
Comment (rudyardrichter) [inline arborist/aws_client.go]: this function not used

File: arborist/aws_client.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+		Region: aws.String(client.credentials.region),
+		Credentials: credentials.NewStaticCredentials(
+			client.credentials.awsAccessKeyID, client.credentials.awsSecretAccessKey, ""),
+	})
+	client.session = sess
+
+	return err
+}
+
+func (client *AwsClient) GetBucketName() string {
```
Comment (rudyardrichter) [inline arborist/aws_client.go]: not used

File: arborist/aws_client.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	return client.bucket
+}
+
+func (client *AwsClient) SetBucketName(name string) {
```
Comment (rudyardrichter) [inline arborist/aws_client.go]: not used

File: arborist/engine.go [hunk 7]
```diff
@@ -660,3 +666,114 @@ func (engine *Engine) listAuthedResources(policyIDs []string) ([]*Resource, erro
...
+	}
+
+	awsClient.LoadCredentialFromConfigFile(cfgPath)
+
+	// Perform download object from s3
+	return awsClient.DownloadObjectFromS3(bucket, modelName, path)
+}
+
+// getRootResource get the root resource
+func (engine *Engine) getRootResource(resources []ResourceJSON) *ResourceJSON {
```
Comment (rudyardrichter) [inline arborist/engine.go]: this function should be unnecessary, just use `engine.rootResource.toJSON()`

Comment (giangbui) [reply arborist/engine.go]: Fixed in bc25eb4ad8ccd44d86b7a24694e3989552e789c7

File: arborist/aws_client.go (hunk 6 above)
Comment (giangbui) [reply arborist/aws_client.go]: Fixed in bc25eb4ad8ccd44d86b7a24694e3989552e789c7

File: arborist/aws_client.go (hunk 5 above)
Comment (giangbui) [reply arborist/aws_client.go]: Fixed in bc25eb4ad8ccd44d86b7a24694e3989552e789c7

File: arborist/aws_client.go (hunk 4 above)
Comment (giangbui) [reply arborist/aws_client.go]: Fixed in bc25eb4ad8ccd44d86b7a24694e3989552e789c7

File: arborist/resource.go (hunk 3 above)
Comment (giangbui) [reply arborist/resource.go]: Fixed in bc25eb4ad8ccd44d86b7a24694e3989552e789c7

File: arborist/server/router_model.go (hunk 1 above)
Comment (giangbui) [reply arborist/server/router_model.go]: Fixed in bc25eb4ad8ccd44d86b7a24694e3989552e789c7

File: arborist/resource_json.go (hunk 2 above)
Comment (giangbui) [reply arborist/resource_json.go]: Fixed in bc25eb4ad8ccd44d86b7a24694e3989552e789c7

**Comment by rudyardrichter** [comment]
out of date

**Pull request closed without merging**